capstone-rs/src/instruction.rs

```diff
@@ -214,53 +171,153 @@ impl<'a> Insn<'a> {
     /// no-way actually tied to the cs_insn itself. It is the responsibility of the caller to
     /// ensure that the resulting `Insn` lives only as long as the `cs_insn`. This function
     /// assumes that the pointer passed is non-null and a valid `cs_insn` pointer.
+    ///
+    /// The caller is fully responsible for the backing allocations lifetime, including freeing.
     pub unsafe fn from_raw(insn: *const cs_insn) -> Self {
         Self {
-            insn: *insn,
+            insn: core::ptr::read(insn),
             _marker: PhantomData,
         }
     }
 
     /// The mnemonic for the instruction
     pub fn mnemonic(&self) -> Option<&str> {
-        unsafe { str_from_cstr_ptr(self.insn.mnemonic.as_ptr()) }
+        unsafe { str_from_cstr_ptr(self.insn().mnemonic.as_ptr()) }
     }
 
     /// The operand string associated with the instruction
     pub fn op_str(&self) -> Option<&str> {
-        unsafe { str_from_cstr_ptr(self.insn.op_str.as_ptr()) }
+        unsafe { str_from_cstr_ptr(self.insn().op_str.as_ptr()) }
     }
 
     /// Access instruction id
     pub fn id(&self) -> InsnId {
-        InsnId(self.insn.id)
+        InsnId(self.insn().id)
     }
 
     /// Size of instruction (in bytes)
-    fn len(&self) -> usize {
-        self.insn.size as usize
+    pub fn len(&self) -> usize {
+        self.insn().size as usize
     }
 
     /// Instruction address
     pub fn address(&self) -> u64 {
-        self.insn.address as u64
+        self.insn().address as u64
     }
 
     /// Byte-level representation of the instruction
     pub fn bytes(&self) -> &[u8] {
-        &self.insn.bytes[..self.len()]
+        &self.insn().bytes[..self.len()]
     }
 
     /// Returns the `Detail` object, if there is one. It is up to the caller to determine
     /// the pre-conditions are satisfied.
     ///
     /// Be careful this is still in early stages and largely untested with various `cs_option` and
     /// architecture matrices
+    ///
+    /// # Safety
+    /// The [`cs_insn::detail`] pointer must be valid and non-null.
     pub(crate) unsafe fn detail(&self, arch: Arch) -> InsnDetail {
-        InsnDetail(&*self.insn.detail, arch)
+        InsnDetail(&*self.insn().detail, arch)
+    }
+
+    // Any pointers are owned by the Capstone C library and should be freed accordingly
+    #[inline(always)]
```

Comment:
Is there a benefit to add an getter method here? I think it may be simpler (and cause fewer lines to change) if you get rid of this method.

Reply:
This was a leftover from implementing `Instruction` as a trait, good catch